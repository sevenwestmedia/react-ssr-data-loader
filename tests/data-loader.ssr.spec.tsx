import { ComponentFixture } from './helpers/component-fixture'
import { SharedDataComponentFixture } from './helpers/shared-data-component-fixture'
import Adapter from 'enzyme-adapter-react-16'
import { configure } from 'enzyme'
import { act } from 'react-dom/test-utils'

configure({ adapter: new Adapter() })

describe('server side render', () => {
    it('should start loading data if not loaded', () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: true })

        sut.assertState()
    })

    it('should pass loaded data once promise resolves', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: true })

        await act(() =>
            sut.testDataPromise.resolve({
                result: 'Success!',
            }),
        )

        sut.assertState()
    })

    it('should pass failure when data load fails', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: true })

        await act(() => sut.testDataPromise.reject(new Error('Boom!')))

        sut.assertState()
    })

    it('second SSR when data loaded should not reload data', async () => {
        let sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: true })
        await act(() => sut.testDataPromise.resolve({ result: 'Success!' }))

        sut = new ComponentFixture(sut.currentState, 'testKey', { isServerSideRender: true })

        expect(sut.loadDataCount).toBe(0)
    })

    it('second SSR when data load failed should not reload data', async () => {
        let sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: true })
        await act(() => sut.testDataPromise.reject(new Error('Boom')))

        sut = new ComponentFixture(sut.currentState, 'testKey', { isServerSideRender: true })

        expect(sut.loadDataCount).toBe(0)
    })

    it('does not trigger data load on the server if clientLoadOnlySet', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', {
            isServerSideRender: true,
            clientLoadOnly: true,
        })

        sut.assertState()
    })

    it('can have two components in the same render tree', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', true)

        sut.assertState()
    })

    it('second SSR with two components in same tree uses data', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', true)
        await act(() => sut.testDataPromise.resolve({ result: 'Success!' }))

        const sut2 = new SharedDataComponentFixture(sut.currentState, 'testKey', true)

        sut2.assertState()
    })
})
