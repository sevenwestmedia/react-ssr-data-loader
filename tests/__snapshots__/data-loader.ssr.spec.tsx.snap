// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server side render can have two components in the same render tree 1`] = `
Object {
  "loadAllCompletedCalled": 0,
  "loadDataCount": 1,
  "renderCount": 1,
  "renderCount2": 1,
  "renderProps1": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
  "renderProps2": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;

exports[`server side render does not render on the server if clientLoadOnlySet 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
  ],
  "loadAllCompletedCalled": 0,
  "loadDataCount": 0,
  "renderActions": undefined,
  "renderCount": 0,
  "renderProps": undefined,
}
`;

exports[`server side render second SSR with two components in same tree uses data 1`] = `
Object {
  "loadAllCompletedCalled": 0,
  "loadDataCount": 0,
  "renderCount": 1,
  "renderCount2": 1,
  "renderProps1": Object {
    "data": Object {
      "dataFromServerSideRender": true,
      "hasData": true,
      "result": Object {
        "result": "Success!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
  "renderProps2": Object {
    "data": Object {
      "dataFromServerSideRender": true,
      "hasData": true,
      "result": Object {
        "result": "Success!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`server side render should pass failure when data load fails 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "error": [Error: Boom!],
            "success": false,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "error": [Error: Boom!],
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "load-error",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
  ],
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "error": [Error: Boom!],
      "success": false,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`server side render should pass loaded data once promise resolves 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": true,
            "hasData": true,
            "result": Object {
              "result": "Success!",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
  ],
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": true,
      "hasData": true,
      "result": Object {
        "result": "Success!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`server side render should start loading data if not loaded 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
  ],
  "loadAllCompletedCalled": 0,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 1,
  "renderProps": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;
